_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,871 @@
+
+{: #cortex-m3-m4-debug-trick}
+For the Cortex M3 & Cortex M4 **only**, there is a trick to make all `IMPRECISE` accesses `PRECISE` by disabling any write buffering. This can be done by setting bit 1 (`DISDEFWBUF`) of the register to 1.
+
+For the Cortex M7, there is **no** way to force all stores to be synchronous.
+
+#### Auxiliary Bus Fault Status Register (ABFSR) - 0xE000EFA8
+
+This register **only** exists for Cortex-M7 devices. When an `IMPRECISE` error occurs it will at least give you an indication of what memory bus the fault occurred on[^7]:
+
+![](img/cortex-m-fault/abfsr.png)
+
+A full discussion of memory interfaces is outside the scope of this article but more details can be found in the reference manual [^7].
+
+{: #mmfsr}
+
+### MemManage Status Register (MMFSR) - 0xE000ED28
+
+This register reports Memory Protection Unit faults. For a deep dive into the peripheral, see [this article](https://interrupt.memfault.com/blog/fix-bugs-and-secure-firmware-with-the-mpu).
+
+> NOTE: For the most part MPU faults will only trigger if the MPU has been enabled. However, there are a few memory access errors that will always result in a MemManage fault -- such as trying to execute from the `0xExxx.xxxx` address range.
+
+The layout of the register looks like this:
+
+![](img/cortex-m-fault/mmfsr.png)
+
+where,
+
+- `MMARVALID` - Indicates the address which triggered the MemManage fault is stored in the _MemManage Fault Address Register_ (**MMFAR**) - a 32 bit register located at `0xE000ED34`
+- `MLSPERR` & `MSTKERR` - Indicates that a MemManage fault occurred during lazy state preservation or exception entry. This could happen if you are using an MPU region as a stack guard in your system
+- `MUNSTKERR` - Indicates that a fault occurred while returning from an exception
+- `DACCVIOL` - Indicates that a data access triggered the MemManage fault.
+- `IACCVIOL` - Indicates that an attempt to execute an instruction triggered a MPU or Execute Never (XN) error.
+
+## HardFault Status Register (HFSR) - 0xE000ED2C
+
+This registers explains the reason a HardFault exception was triggered.
+
+![](img/cortex-m-fault/hfsr.png)
+
+There's not too much information in this register but we will go over the fields real quickly
+
+- `DEBUGEVT` - Indicates that a debug event (i.e executing a breakpoint instruction) occurred while the debug subsystem was not enabled
+- `FORCED` - This means a fault of configurable priority, the faults we just discussed [above](#cfsr) has been escalated to a hardfault. The exception handlers for configurable faults are not enabled by default so this is pretty common. If the exception handlers are enabled, a configurable fault may still escalate to a hardfault if the configurable fault occurred at a higher priority than the exception handler
```

Comment:
> This means a fault of configurable priority, the faults we just discussed [above](#cfsr) has been escalated to a hardfault

This sentence could be simpler. What about "A configurable fault (i.e. the fault types we discussed in previous sections) was escalated to a hardfault".